### ISSUE TITLE
Docker continues writing to rotated container log files

### ISSUE THREAD

**Issue by kunickiaj** [CiscoCloud/microservices-infrastructure#500]
First, I'd appreciate if someone could verify they see the same behavior.

It appears that once the daily logrotate of the docker container logs runs, docker continues to write logs to the rotated file instead of the current log. This is the behavior whether or not logrotate is configured with `copytruncate`.

e.g.
```
-rw------- 1 root root           0 May 30 03:33 ed5c737c96c0f50ee5632ff28634da64276daaa659bdcae8fbfa4331598f4edc/ed5c737c96c0f50ee5632ff28634da64276daaa659bdcae8fbfa4331598f4edc-json.log
-rw------- 1 root root  1981069563 Jun 17 16:23 ed5c737c96c0f50ee5632ff28634da64276daaa659bdcae8fbfa4331598f4edc/ed5c737c96c0f50ee5632ff28634da64276daaa659bdcae8fbfa4331598f4edc-json.log-20150530
```

```
-rw------- 1 root root          0 May 30 03:36 e28ddc4a010b6878d74b687eb43e9a2241ac12949180982bc4868344f2b313c2/e28ddc4a010b6878d74b687eb43e9a2241ac12949180982bc4868344f2b313c2-json.log
-rw------- 1 root root 1576986207 Jun 17 16:24 e28ddc4a010b6878d74b687eb43e9a2241ac12949180982bc4868344f2b313c2/e28ddc4a010b6878d74b687eb43e9a2241ac12949180982bc4868344f2b313c2-json.log-20150530
```

In both cases a container was started on May 30, the log was rotated that night, and the container continues to write to the May 30 log file instead of the rotated one.

**Comment by stevendborrelli** [comment]
Thanks for brining this to our attention. We're investigating.

**Comment by SillyMoo** [comment]
This is a known issue: https://github.com/docker/docker/issues/7333. The underlying problem is that there is no way of signalling to docker to stop writing to a file and start a new one. This means when logrotate moves the file the FD is still the same and so it just carries on writing to the now moved file.
The suggestion seems to be that you need to use the syslog driver, but it looks like this needs to be configured on a per-container basis which is obviously not idea.

**Comment by SillyMoo** [comment]
I tell a lie, you can use the syslog driver as an option when starting the docker daemon (-log-driver="syslog"). There is also support for fluentd, gelf (logstash) and journalctl drivers.

**Comment by altvnk** [comment]
Seems like fluentd is a good option, I saw some writings in their blog. Btw fluentd itself has a good performance and smaller footprint then logstash. Also primarily fluentd is log shipper.

**Comment by SillyMoo** [comment]
I figure a good short term fix is to use syslog+logrotate (as it's not far off what is there already) and maybe logstash later (as it looks like there is plenty of enhancements mentioning logstash). If the project owners are ok with it I'm happy to take on this issue and put together a pull request for the syslog+logrotate solution.

**Comment by SillyMoo** [comment]
Created a pull request that does the syslog based immediate solution, tested it with logrotate --force and seems to work. Puts logs in /var/log/docker/docker.log, and archives them (as per mesos) to /var/log/docker/archive.

**Comment by ryane** [comment]
Merged #648 with docker syslog driver to address this issue for now. More discussion on improving logging infrastructure happening in #646
